File: benches/moat/governance_lines/apps/refund_bot/typescript/refund_bot.ts
// Refund-bot reference implementation — TypeScript.
//
// Every governance check is application code: a runtime guard, a zod
// schema validation, a manual audit-log push, a `sources: string[]`
// field threaded through every call. tsc enforces shape, not policy.
//
// Lines marked with `// governance` are the ones the counter classifies.

import { z } from "zod"; // governance

// governance: in-memory audit log; convention-only.
const auditLog: Array<Record<string, unknown>> = []; // governance

const TrustLevel = z.enum([ // governance
  "autonomous", // governance
  "supervisor", // governance
  "human_required", // governance
]); // governance
type TrustLevel = z.infer<typeof TrustLevel>; // governance

const RefundRequest = z.object({
  order_id: z.string(),
  amount: z.number().positive().max(500), // governance (budget cap)
  reason: z.string(),
});
type RefundRequest = z.infer<typeof RefundRequest>;

const RefundResponse = z.object({
  receipt_id: z.string(),
  status: z.string(),
});
type RefundResponse = z.infer<typeof RefundResponse>;

const RefundExplanation = z.object({
  reason: z.string(),
  sources: z.array(z.string()).default([]), // governance (provenance)
});
type RefundExplanation = z.infer<typeof RefundExplanation>;

interface Approval { // governance
  trust: TrustLevel; // governance
  actor: string; // governance
} // governance

function dangerous<TArgs extends unknown[], TRet>( // governance
  required: TrustLevel, // governance
  fn: (...args: TArgs) => TRet, // governance
): (approval: Approval, ...args: TArgs) => TRet { // governance
  return (approval, ...args) => { // governance
    if (!approval) { // governance
      throw new Error("dangerous tool requires an approval token"); // governance
    } // governance
    if (approval.trust !== required) { // governance
      throw new Error("approval trust level mismatch"); // governance
    } // governance
    auditLog.push({ // governance
      tool: fn.name, // governance
      approval, // governance
      ts: Date.now(), // governance
    }); // governance
    return fn(...args); // governance
  }; // governance
} // governance

const issue_refund = dangerous("human_required", function issue_refund(req: RefundRequest): string { // governance (decorator wrap)
  return `r-${req.order_id}`;
});

function fetch_order(order_id: string): { text: string; sources: string[] } { // governance (return shape)
  // governance: must return text + sources so callers can preserve citations.
  return { // governance
    text: `order ${order_id} placed 2026-04-21`, // governance
    sources: [`db://orders/${order_id}`], // governance
  }; // governance
}

export function approve_refund(req: RefundRequest): RefundResponse {
  const approval: Approval = { trust: "human_required", actor: "human" }; // governance
  const receipt_id = issue_refund(approval, req);
  return { receipt_id, status: "approved" };
}

export function explain_refund(order_id: string): RefundExplanation {
  const { text, sources } = fetch_order(order_id);
  if (sources.length === 0) { // governance (catches dropped citation)
    throw new Error("ungrounded explanation rejected"); // governance
  } // governance
  return { reason: text, sources };
}

File: benches/moat/governance_lines/apps/refund_bot/python/refund_bot.py
"""Refund-bot reference implementation — Python.

Every governance check in this file is *application code*: an `if`
statement, an enum lookup, a custom decorator, a runtime audit-log
write, a manual citation list. The type system has nothing to enforce;
removing any of these lines lets a real bug ship.

Lines tagged `# governance` in trailing comments are the ones the
counter classifies. The pattern is intentional — a Python developer
who removes a `# governance` line ships a real bug class.
"""

from __future__ import annotations

import functools
import logging
import time
from dataclasses import dataclass, field
from enum import Enum
from typing import Annotated, Callable, TypeVar

from pydantic import BaseModel, Field, ValidationError  # governance

# governance: trace event sink. Convention-only.
_AUDIT_LOG: list[dict[str, object]] = []  # governance


class TrustLevel(str, Enum):  # governance
    AUTONOMOUS = "autonomous"
    SUPERVISOR = "supervisor"
    HUMAN_REQUIRED = "human_required"


class RefundRequest(BaseModel):
    order_id: str
    amount: Annotated[float, Field(gt=0, le=500.0)]  # governance (budget cap)
    reason: str


class RefundResponse(BaseModel):
    receipt_id: str
    status: str


class RefundExplanation(BaseModel):
    reason: str
    sources: list[str] = field(default_factory=list)  # governance (provenance)


F = TypeVar("F", bound=Callable[..., object])


def dangerous(trust: TrustLevel) -> Callable[[F], F]:  # governance
    """Decorator marking a tool dangerous. Convention-only enforcement."""

    def decorator(fn: F) -> F:
        @functools.wraps(fn)
        def wrapper(*args: object, **kwargs: object) -> object:
            approval = kwargs.pop("approval", None)  # governance
            if approval is None:  # governance
                raise PermissionError(  # governance
                    f"{fn.__name__} requires an approval token"  # governance
                )  # governance
            if approval.get("trust") != trust.value:  # governance
                raise PermissionError(  # governance
                    f"approval trust level mismatch for {fn.__name__}"  # governance
                )  # governance
            _AUDIT_LOG.append(  # governance
                {  # governance
                    "tool": fn.__name__,  # governance
                    "approval": approval,  # governance
                    "ts": time.time(),  # governance
                }  # governance
            )  # governance
            return fn(*args, **kwargs)

        return wrapper  # type: ignore[return-value]

    return decorator


@dangerous(TrustLevel.HUMAN_REQUIRED)  # governance
def issue_refund(req: RefundRequest) -> str:
    return f"r-{req.order_id}"


def fetch_order(order_id: str) -> tuple[str, list[str]]:  # governance (return shape)
    # governance: must return a (text, sources) pair so callers can preserve
    # citations. Forgetting the second element is a real bug class.
    text = f"order {order_id} placed 2026-04-21"
    sources = [f"db://orders/{order_id}"]  # governance
    return text, sources  # governance


def approve_refund(req: RefundRequest) -> RefundResponse:
    # governance: caller must construct the approval token correctly.
    approval = {"trust": TrustLevel.HUMAN_REQUIRED.value, "actor": "human"}  # governance
    receipt_id = issue_refund(req, approval=approval)  # type: ignore[call-arg]
    return RefundResponse(receipt_id=str(receipt_id), status="approved")


def explain_refund(order_id: str) -> RefundExplanation:
    text, sources = fetch_order(order_id)
    if not sources:  # governance (catches dropped citation)
        raise ValueError("ungrounded explanation rejected")  # governance
    return RefundExplanation(reason=text, sources=sources)
